### ISSUE TITLE
Reporting initial status time: 20.890889 seconds.

### ISSUE THREAD

**Issue by jpopelka** [packit/packit-service#1728]
Yesterday, I noticed on staging several cases of the initial status time being set after 15 seconds, which is the limit for the [SLO1 error budget](https://github.com/packit/research/blob/main/error-budgets/SLOs-SLIs.md#slo1-changes-to-github-prs-receive-a-status-update-within-15-seconds-in-99-of-cases).

Looking at the short-running worker logs, you can see that there was a window of 17 seconds between the `process_message` parsed the event and started to get the package config where (as it looks to me) it was just slacking.
```
[12:34:19,375: INFO/MainProcess] Task task.steve_jobs.process_message[e4622d3e-786a-4821-bfb2-f0069bf63fa8] received
[12:34:19,376: DEBUG/MainProcess] TaskPool: Apply <function fast_trace_task at 0x7f7aa5d28af0> (args:('task.steve_jobs.process_message', 'e4622d3e-786a-4821-bfb2-f0069bf63fa8',... )
[12:34:19,377: INFO/MainProcess] Github PR#2 comment: '/packit-stg build' id#1291965878 'created' event.
[12:34:19,378: INFO/MainProcess] Target repo: lbarcziova/hello-world.
[12:34:20,967: DEBUG/MainProcess] pidbox received method enable_events() [reply_to:None ticket:None]
[12:34:25,968: DEBUG/MainProcess] pidbox received method enable_events() [reply_to:None ticket:None]
[12:34:30,968: DEBUG/MainProcess] pidbox received method enable_events() [reply_to:None ticket:None]
[12:34:35,968: DEBUG/MainProcess] pidbox received method enable_events() [reply_to:None ticket:None]
[12:34:36,079: DEBUG/MainProcess] Getting package_config: ...
[12:34:36,646: DEBUG/MainProcess] Found a config file '.packit.yaml' on ref '...' of the 'lbarcziova/hello-world' repository.
...
[12:34:40,704: DEBUG/MainProcess] Reporting initial status time: 20.890889 seconds.
...
[12:34:44,298: INFO/MainProcess] Task task.steve_jobs.process_message[e4622d3e-786a-4821-bfb2-f0069bf63fa8] succeeded in 24.920801570871845s: ...
```

When checking the `Average Task Runtime at Worker` graph in the `Packit boards / Celery Monitoring with Flower` 
I can see that the average task runtime of `steve_jobs.process_message` is from (min) units of milliseconds to (max) approx 15 seconds.

Shouldn't the `process_message` be just some pre-processing, like parsing, pre-checking, getting handlers, and creating tasks?
Is there something else, like some networking I/O which can take a long time (>10s)?
Maybe adding more debug logs so that we can narrow this down when it happens again?

Related to #1677, #1364

**Comment by lachmanfrantisek** [comment]
> Is there something else, like some networking I/O which can take a long time (>10s)?

I think we are doing also pre-checks where we e.g. verify the permissions (db + forge API calls). We also set statuses for various deprecations/problems/... We can maybe move that (or as much as possible) to the job execution itself. (What probably means getting rid of that `pre-check` completely.)

**Comment by TomasTomecek** [comment]
It would be so helpful if we knew how many network requests, I/O operations and SQL queries happen and how long they take. In theory, we could do something like:
```
if initial_status_reported_after > time_delta("15s"):
  logger.error("We have failed SLO1 for this event, let's get it to sentry so we can analyze further")
```
sentry contains logs from other modules so maybe there will be something in there

**Comment by majamassarini** [comment]
The [handler's refactoring PR](https://github.com/packit/packit-service/pull/1678) was merged in stage the 25th of October. 
And the logs seems to be referring the stage instance at that time.

The linked PR is for sure getting performance worst.
Before the refactoring many objects (ServiceConfig, GitProject, PackitAPIProtocol, Helper classes ecc.) were initialized just once for every handler, now those objects are initialized once for every handler and once for every checker in the handler.

I can probably say that now we are at least twice slower than before at doing the same check.

**Comment by jpopelka** [comment]
I once more reverted (at the time of writing this) the packit-worker-short-running@prod from 1 multi-threaded instance to 4 single-threaded/processed instances so that:
- it's easier to check the logs (because all the threads write into one log)
- to rule out the multi-threaded worker as a cause of the worsened performance (#1677)

I plan to keep it like this for at least several weeks to make some conclusion because [last time](https://github.com/packit/packit-service/issues/1677#issuecomment-1292199866) I kept it for 2 weeks and it was hard to tell whether there was a change of trend.

**Comment by TomasTomecek** [comment]
We are going to run out of error budget in the coming days:

![image](https://user-images.githubusercontent.com/1662493/202165320-7fd63899-5027-4bf0-b2b5-6103f0603a08.png)

Those big spikes look really concerning, it's also interesting they are clustered like that - as if something has happened in that timeframe that caused the failures

![image](https://user-images.githubusercontent.com/1662493/202166169-a28f386f-f132-4158-a33d-9f812a5cba83.png)

**Comment by TomasTomecek** [comment]
Based on the sentry issue above, I think the way we measure SLO1 right now is incorrect.

You can browse the logs above and see that every GitHub `POST /check-runs` request takes ~1 second. In the measurement of the response time, we include setting all "accepted" and "queued" status updates.

I'd say we should measure up to setting the first "accepted" state.

```
0:00 tf-f-rawhide - accepted  # expected
0:01 tf-f-f37 - accepted
0:02 tf-f-f36 - accepted
0:05  # Packit doing packit things
0:05 tf-f-rawhide - queued
0:06 tf-f-f37 - queued
0:07 tf-f-f36 - queued  # now
```

**Comment by jpopelka** [comment]
The Sentry's breadcrumbs are amazing. I agree that it's clear from them that setting the checks take the most time (usually about 1s, but sometimes I can see up to 5s).

The problem, in my opinion, however, is that the task that processes the incoming event sets all the "queued" (I don't see any "accepted" statuses Tomas) checks for all targets - and if there are more (like 8-10) targets, then just setting the checks could easily take more than 15s.

My idea is that if the initial task only created the follow-up tasks and let them set the "queued" checks, it might actually be faster because the follow-up tasks are processes in parallel (by more parallel threads or individual workers).

Or have a task whose only purpose was to set the ("queued") state. And the initial task would create a follow-up task + the set-check task for each target.

**Comment by TomasTomecek** [comment]
Thank you Jirka for your detailed explanation.

I think I understand your PoV. I think :D

How about we split SLO1 into SLO1.1 and SLO1.2:

SLO1.1 - 1st status set no less than 15s
SLO1.2 - all statues set within 40 seconds (arbitrary number)

That way we know that Packit reacted to a PR and all jobs are set up in a reasonable time (that's what I understand from your explanation).

The reason I propose this is that our budget for both .1 and 2. should be 100% all the time and would decrease actually if something goes wrong.

Let's talk on Monday :)

**Comment by TomasTomecek** [comment]
We agreed to do ^

okay so this ticket is trivial then:
* [ ] set the threshold from 15s to 40s (still an arbitrary number)
* [ ] update docs in packit.dev
* [ ] and our grafana dashboard
